Add new rule to check if null checking is being done using short circuit (#238)
terraform-linters/tflint-ruleset-terraform: docs/rules/terraform_no_short_circuit_evaluation.md

> # terraform_no_short_circuit_evaluation

Reviewer: `terraform_logical_short_circuit` would be a better name. Most of the other rules are anti-patterns but none uses a `no` prefix.

> Disallow using logical operators (`&&`, `||`) with null checks that could lead to errors due to lack of short-circuit evaluation.

Reviewer: ```suggestion
Detects usage of logical operators (`&&`, `||`) that could lead to errors because of their lack of support for short-circuit evaluation.
```

Doesn't seem like null is an essential part of the description, you might be looking at length or any other intended pre-condition.

> This rule is enabled by "recommended" preset.

## Example

```hcl
# This will error if var.obj is null
resource "aws_instance" "example" {
  count = var.obj != null && var.obj.enabled ? 1 : 0
}

# This is the safe way to write it
resource "aws_instance" "example" {
  count = var.obj != null ? var.obj.enabled ? 1 : 0 : 0
}
```

```
$ tflint
1 issue(s) found:

Warning: Short-circuit evaluation is not supported in Terraform. Use a conditional expression (condition ? true : false) instead. (terraform_no_short_circuit_evaluation)

  on main.tf line 3:
   3:   count = var.obj != null && var.obj.enabled ? 1 : 0

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.1.0/docs/rules/terraform_no_short_circuit_evaluation.md
```

## Why

Unlike many programming languages, Terraform's logical operators (`&&` and `||`) do not short-circuit. This means that in an expression like `var.obj != null && var.obj.enabled`, both sides will be evaluated even if `var.obj` is null, which will result in an error.

> This is a common source of confusion for users coming from other programming languages where short-circuit evaluation is standard behavior. The issue is particularly problematic when checking for null before accessing object attributes.

Reviewer: > particularly problematic

Compared to what?

## How To Fix

Use nested conditional expressions instead of logical operators when you need short-circuit behavior. For example:

> ```hcl

Reviewer: Split this into multiple code blocks for readability

# Instead of this:
var.obj != null && var.obj.enabled

# Use this:
var.obj != null ? var.obj.enabled : false

# For more complex conditions:
var.obj != null ? (var.obj.enabled ? var.obj.value > 0 : false) : false
```

You can also use the `try()` function in some cases, though this may mask errors you want to catch:

```hcl
try(var.obj.enabled, false)
```

> For more information, see [hashicorp/terraform#24128](https://github.com/hashicorp/terraform/issues/24128). 

Reviewer: This is too open-ended for users, too much history to sort through. Instead link to the official docs on this limitation:

https://developer.hashicorp.com/terraform/language/expressions/operators#logical-operators

And then perhaps summarize any version-specific context, e.g. if it only affects Terraform versions earlier than _x_.
